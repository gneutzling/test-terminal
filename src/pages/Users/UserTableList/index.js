import React from "react"
import { Link as RouterLink } from "react-router-dom"
import Button from "@material-ui/core/Button"
import Link from "@material-ui/core/Link"
import Table from "@material-ui/core/Table"
import TableBody from "@material-ui/core/TableBody"
import TableCell from "@material-ui/core/TableCell"
import TableHead from "@material-ui/core/TableHead"
import TableRow from "@material-ui/core/TableRow"
import { getETHBalance } from "app/utils/getETHBalance"

const UserTableList = props => {
  return (
    <Table>
      <TableHead>
        <TableRow>
          <TableCell>User ID</TableCell>
          <TableCell colSpan={2}>ETH Balance</TableCell>
        </TableRow>
      </TableHead>
      <TableBody>
        {props.users.map(user => {
          return (
            <TableRow key={user.id}>
              <TableCell>{user.id}</TableCell>
              <TableCell>{getETHBalance(user.exchangeBalances)}</TableCell>
              <TableCell>
                <Button color="primary">
                  <Link
                    component={RouterLink}
                    underline="none"
                    to={`/users/${user.id}`}
                  >
                    See transactions
                  </Link>
                </Button>
              </TableCell>
            </TableRow>
          )
        })}
      </TableBody>
    </Table>
  )
}

export default UserTableList
