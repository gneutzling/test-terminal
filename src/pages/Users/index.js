import React from "react"
import { pathOr, pipe, pluck, subtract, sum } from "ramda"
import { Redirect, Route, Switch, Link } from "react-router-dom"
import UserTransactions from "app/pages/UserTransactions"
import Query from "app/common/Query"
import QUERY_USERS from "./query"

const Users = () => {
  const getUserETHBalance = exchangeBalances => {
    const ethBoughtTotal = pipe(
      pluck("ethBought"),
      sum
    )(exchangeBalances)

    const ethSoldTotal = pipe(
      pluck("ethSold"),
      sum
    )(exchangeBalances)

    return subtract(ethBoughtTotal, ethSoldTotal)
  }

  return (
    <div>
      <h1>User list</h1>

      <Query query={QUERY_USERS}>
        {({ data }) => {
          const users = pathOr([], ["users"], data)

          return (
            <div>
              {users.map(user => {
                return (
                  <div key={user.id}>
                    <p>ID: {user.id}</p>
                    <p>
                      ETH Balance: {getUserETHBalance(user.exchangeBalances)}
                    </p>
                    <Link to={`/users/${user.id}`}>See transactions</Link>
                    <hr />
                  </div>
                )
              })}
            </div>
          )
        }}
      </Query>

      <Switch>
        <Route
          exact
          path="/users/:userId"
          render={({ match }) => {
            return (
              <Redirect to={`/users/${match.params.userId}/transactions`} />
            )
          }}
        />
        <Route
          path="/users/:userId/transactions"
          component={UserTransactions}
        />
      </Switch>
    </div>
  )
}

export default Users
