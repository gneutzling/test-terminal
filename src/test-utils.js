import React from "react"
import { render } from "@testing-library/react"
import "jest-dom/extend-expect"

import RouterProvider from "app/lib/RouterProvider"

const AllTheProviders = ({ children }) => {
  return <RouterProvider>{children}</RouterProvider>
}

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options })

// re-export everything
export * from "@testing-library/react"

// override render method
export { customRender as render }
