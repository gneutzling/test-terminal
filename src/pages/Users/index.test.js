import React from "react"
import { MockedProvider } from "react-apollo/test-utils"
import wait from "waait"
import { render } from "app/test-utils"
import QUERY_USERS from "./query"
import Users from "./"

const build = props =>
  render(
    <MockedProvider mocks={props.mocks} addTypename={false}>
      <Users />
    </MockedProvider>
  )

const defaultMocks = [
  {
    request: {
      query: QUERY_USERS,
    },
    result: {
      data: {
        users: [
          {
            id: "0x0000000000000000000000000000000000000000",
            exchangeBalances: [],
          },
          {
            id: "0x0000000000c90bc353314b6911180ed7e06019a9",
            exchangeBalances: [
              {
                id:
                  "0x0cf0ee63788a0849fe5297f3407f701e122cc023-0x0000000000c90bc353314b6911180ed7e06019a9",
                ethBought: "0",
                ethSold: "10",
              },
              {
                id:
                  "0xe7775a6e9bcf904eb39da2b68c5efb4f9360e08c-0x0000000000c90bc353314b6911180ed7e06019a9",
                ethBought: "40",
                ethSold: "10",
              },
            ],
          },
        ],
      },
    },
  },
]

describe("Users", () => {
  it("matches snapshot", async () => {
    const { container } = build({ mocks: defaultMocks })
    await wait(0) // wait for response
    expect(container).toMatchSnapshot()
  })
})
