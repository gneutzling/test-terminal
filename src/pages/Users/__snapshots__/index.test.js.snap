// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users matches snapshot 1`] = `
<div>
  <div>
    <h1>
      User list
    </h1>
    <div>
      <div>
        <p>
          ID: 
          0x0000000000000000000000000000000000000000
        </p>
        <p>
          ETH Balance: 
          0
        </p>
        <a
          href="/users/0x0000000000000000000000000000000000000000"
        >
          See transactions
        </a>
        <hr />
      </div>
      <div>
        <p>
          ID: 
          0x0000000000c90bc353314b6911180ed7e06019a9
        </p>
        <p>
          ETH Balance: 
          20
        </p>
        <a
          href="/users/0x0000000000c90bc353314b6911180ed7e06019a9"
        >
          See transactions
        </a>
        <hr />
      </div>
    </div>
  </div>
</div>
`;
